### PR TITLE
Add `torch._C._set_watchdog()` that spawns a pthread-watchdog

### PR DESCRIPTION
**Pull request by malfet**
Hopefully that should be reliable enough to detect any native/managed deadlocks while running python testsuite

### CODE REVIEW

File: torch/csrc/Module.cpp [hunk 1]
```diff
@@ -524,6 +529,58 @@ PyObject *THPModule_isEnabledXNNPACK(PyObject * /* unused */)
...
+    return nullptr;
+  }
+  std::string name(cname);
+  auto it = watchdogs.find(name);
+  if (it != watchdogs.end()) {
+   if (timeout == 0) {
+     watchdogs.erase(it);
+     Py_RETURN_NONE;
+   }
+   std::cerr << " Ignoring watchdog override " << name << std::endl;
```
Comment (ezyang) [inline torch/csrc/Module.cpp:577]: lol

Comment (ezyang) [reply torch/csrc/Module.cpp:577]: Can we just make this a hard error?

File: torch/csrc/Module.cpp [hunk 2]
```diff
@@ -524,6 +529,58 @@ PyObject *THPModule_isEnabledXNNPACK(PyObject * /* unused */)
...
 
+namespace {
+struct WatchdogThread {
+  WatchdogThread(const std::string& name_, int timeout): name(name_), thrd([this, timeout] {
+      using namespace std::chrono_literals;
+      std::unique_lock<std::mutex> lock(mutex);
+      cv.wait_for(lock, timeout *1s, [this] { return exitFlag; });
+      if (!exitFlag) {
+        std::cerr << "Test " << name << " takes longer than " << timeout << " seconds" <<std::endl;
+        std::exit(-1);
```
Comment (ezyang) [inline torch/csrc/Module.cpp:540]: Are we sure we want a `std::exit` here? Maybe a `std::abort` would be better at actually killing the process?

File: torch/csrc/Module.cpp [hunk 3]
```diff
@@ -524,6 +529,58 @@ PyObject *THPModule_isEnabledXNNPACK(PyObject * /* unused */)
   else Py_RETURN_FALSE;
 }
 
+namespace {
+struct WatchdogThread {
+  WatchdogThread(const std::string& name_, int timeout): name(name_), thrd([this, timeout] {
+      using namespace std::chrono_literals;
+      std::unique_lock<std::mutex> lock(mutex);
+      cv.wait_for(lock, timeout *1s, [this] { return exitFlag; });
```
Comment (ezyang) [inline torch/csrc/Module.cpp:537]: Second suffix, so fancy!!

File: torch/csrc/Module.cpp [hunk 4]
```diff
@@ -524,6 +529,58 @@ PyObject *THPModule_isEnabledXNNPACK(PyObject * /* unused */)
...
+    cv.notify_one();
+    if (thrd.joinable())
+      thrd.join();
+  }
+private:
+  std::string name;
+  std::mutex mutex;
+  std::condition_variable cv;
+  bool exitFlag = false;
+  // Thread must be the last member, otherwise it will not get properly initialized
```
Comment (ezyang) [inline torch/csrc/Module.cpp:557]: because you need the cv/mutex etc to be initialized before it, is that right? Subtle...

**Review by ezyang — APPROVED** [review]

File: torch/csrc/Module.cpp (hunk 3 above)
Comment (malfet) [reply torch/csrc/Module.cpp:537]: I like `chrono_literls` :)

File: torch/csrc/Module.cpp (hunk 4 above)
Comment (malfet) [reply torch/csrc/Module.cpp:557]: Yes, both `mutex`, `cv` and `exitFlag` must be initialized before thread is created. (I knew about this, but actually it was a bug in my initial implementation and this race was hit half way thru `torch_jit` tests)

File: torch/csrc/Module.cpp (hunk 1 above)
Comment (malfet) [reply torch/csrc/Module.cpp:577]: Don't have strong preference, one way or the other, will add `std::abort()` and see what will happen.

**Pull request closed without merging**